c2corg_api/views/document.py

```diff
@@ -296,6 +296,35 @@ def _put(
 
         return {}
 
+    def _delete(self, clazz, schema, clazz_locale=None):
+        if not self.request.has_permission('moderator'):
+            raise HTTPForbidden('No permission to delete the document')
+        id = self.request.validated['id']
+
+        # get the current version of the document
+        document = self._get_document(clazz, id, clazz_locale=clazz_locale)
+        lang = document.locales[0].lang
+
+        associations = doc_associations.get_associations(document, lang, False)
```

Comment:
You are loading all associations to check if a waypoint is still the main waypoint of a route? This could be improved. You could directly query the routes table:

```
def _is_main_waypoint_of_route(waypoint):
    return DBSession.query(
        exists().
        where(Route.main_waypoint_id == waypoint.document_id)
    ).scalar()
```
Similar to https://github.com/c2corg/v6_api/blob/master/c2corg_api/views/association.py#L161